File: frontend/src/services/api/notifications.ts
import { apiClient } from './client';

export interface Notification {
  id: number;
  user_id: number;
  message: string;
  type: 'info' | 'success' | 'warning' | 'error';
  is_read: boolean;
  created_at: string;
}

export const notificationService = {
  getNotifications: async (skip: number = 0, limit: number = 100): Promise<Notification[]> => {
    return apiClient<Notification[]>(`/notifications/?skip=${skip}&limit=${limit}`);
  },

  markAsRead: async (id: number): Promise<Notification> => {
    return apiClient<Notification>(`/notifications/${id}/read`, {
      method: 'PUT'
    });
  },

  markAllAsRead: async (): Promise<{ message: string }> => {
    return apiClient<{ message: string }>('/notifications/read-all', {
      method: 'PUT'
    });
  }
};

File: backend/app/api/routes/notifications.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from sqlalchemy import update, delete
from typing import List, Any
from datetime import datetime, timedelta

from app.db.database import get_db
from app.models.schema import Notification, User
from app.schemas.notification import NotificationResponse
from app.api.deps import get_current_active_user
from app.core.logger import get_logger

logger = get_logger(__name__)
router = APIRouter()

@router.get("/", response_model=List[NotificationResponse])
async def get_notifications(
    skip: int = 0,
    limit: int = 100,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
) -> Any:
    """
    Retrieve notifications for the current user.
    """
    result = await db.execute(
        select(Notification)
        .where(Notification.user_id == current_user.id)
        .order_by(Notification.created_at.desc())
        .offset(skip)
        .limit(limit)
    )
    return result.scalars().all()

@router.put("/{notification_id}/read", response_model=NotificationResponse)
async def mark_notification_read(
    notification_id: int,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
) -> Any:
    """
    Mark a specific notification as read.
    """
    result = await db.execute(
        select(Notification)
        .where(Notification.id == notification_id, Notification.user_id == current_user.id)
    )
    notification = result.scalars().first()
    
    if not notification:
        raise HTTPException(status_code=404, detail="Notification not found")
        
    notification.is_read = True
    db.add(notification)
    await db.commit()
    await db.refresh(notification)
    
    return notification

@router.put("/read-all", status_code=200)
async def mark_all_notifications_read(
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
) -> Any:
    """
    Mark all notifications for the current user as read.
    """
    await db.execute(
        update(Notification)
        .where(Notification.user_id == current_user.id)
        .values(is_read=True)
    )
    await db.commit()
    return {"message": "All notifications marked as read"}
